Source/Error.swift

```diff
@@ -24,41 +24,540 @@
 
 import Foundation
 
-/// The domain used for creating all Alamofire errors.
-public let ErrorDomain = "org.alamofire.error"
-
-/// The custom error codes generated by Alamofire.
-public enum ErrorCode: Int {
-    case inputStreamReadFailed           = -6000
-    case outputStreamWriteFailed         = -6001
-    case contentTypeValidationFailed     = -6002
-    case statusCodeValidationFailed      = -6003
-    case dataSerializationFailed         = -6004
-    case stringSerializationFailed       = -6005
-    case jsonSerializationFailed         = -6006
-    case propertyListSerializationFailed = -6007
+/// `AFError` is the error type returned by the Alamofire framework. It encompasses a few
+/// different types of errors, each with their own associated reasons.
+///
+/// `.multipartEncodingFailed` errors are returned when some step in the multipart encoding process fails.
+///
+/// `.responseValidationFailed` errors are returned when a `validate()` call fails.
+///
+/// `.responseSerializationFailed` errors are returned when a response serializer encounters an error in the serialization process.
+public enum AFError: Error {
+
+    public enum MultipartEncodingFailureReason {
+        case bodyPartURLInvalid(at: URL)
+        case bodyPartFilenameInvalid(in: URL)
+        case bodyPartFileNotReachable(at: URL)
+        case bodyPartFileNotReachableWithError(url: URL, error: Error)
+        case bodyPartFileIsDirectory(at: URL)
+        case bodyPartFileSizeNotAvailable(at: URL)
+        case bodyPartFileSizeQueryFailedWithError(url: URL, error: Error)
+        case bodyPartInputStreamCreationFailed(for: URL)
+
+        case outputStreamCreationFailed(for: URL)
+        case outputStreamFileAlreadyExists(at: URL)
+        case outputStreamURLInvalid(url: URL)
+        case outputStreamWriteFailed(error: Error)
+
+        case inputStreamReadFailed(error: Error)
+    }
+
+    public enum ValidationFailureReason {
+        case missingContentType(acceptable: [String])
+        case unacceptableContentType(acceptable: [String], response: String)
```

Comment:
These could both probably benefit from more verbosity. What about `acceptableContentTypes` and `responseContentType`?